modules/mobile/app/services/mobile/v0/profile/sync_update_service.rb

```diff
@@ -0,0 +1,136 @@
+      #   service = Mobile::V0::Profile::SyncUpdateService.new(user)
+      #   response = service.save_and_await_response('address', update_params)
+      #
+      class SyncUpdateService
+        TRANSACTION_RECEIVED = 'RECEIVED'
+        TIMEOUT_SECONDS = 55
+
+        def initialize(user)
+          @user = user
+        end
+
+        # Kicks off an update and polls to check for a complete response.
+        # This method will raise a timeout if it hit 10 retries or the 55s
+        # limit (whichever comes first).
+        #
+        # @resource_type String the resource type to update
+        # @params Hash the new parameters used in the update
+        # @update whether the save is a new write or an update
+        #
+        # @return AsyncTransaction::Vet360::Base the final async transaction status
+        #
+        def save_and_await_response(resource_type:, params:, update: true)
+          http_method = update ? 'put' : 'post'
+          initial_transaction = save!(http_method, resource_type, params)
+
+          # return non-received status transactions (errors)
+          return initial_transaction unless initial_transaction.transaction_status == TRANSACTION_RECEIVED
+
+          poll_with_backoff do
+            check_transaction_status!(initial_transaction.transaction_id)
+          end
+        end
+
+        private
+
+        def save!(http_method, resource_type, params)
+          record = build_record(resource_type, params)
+          raise Common::Exceptions::ValidationErrors, record unless record.valid?
+
+          response = contact_information_service.send("#{http_method}_#{resource_type.downcase}", record)
+          "AsyncTransaction::Vet360::#{resource_type.capitalize}Transaction".constantize.start(@user, response)
+        end
+
+        def build_record(type, params)
+          "Vet360::Models::#{type.capitalize}"
+            .constantize
+            .new(params)
+            .set_defaults(@user)
+        end
+
+        def poll_with_backoff
+          try = 0
+          start = seconds_since_epoch
+          begin
+            yield
+          rescue IncompleteTransaction
+            # tries 5 times over the first roughly five seconds, then five more times over the next 50s
+            next_try_seconds = Float(2**try) / 10
+            elapsed = get_elapsed(start)
+            log_incomplete(elapsed, next_try_seconds, try)
+
+            # raise gateway timeout if we're at try number 10 or if the next retry would fall outside the timeout window
+            raise_timeout_error(elapsed, try) if tries_or_time_exhausted?(next_try_seconds, elapsed, try)
+
+            sleep next_try_seconds
+            try += 1
+            retry
+          end
+        end
+
+        def get_elapsed(start)
+          seconds_since_epoch - start
+        end
+
+        def tries_or_time_exhausted?(sleep_duration, elapsed, try)
+          try == 9 || elapsed + sleep_duration > TIMEOUT_SECONDS
```

Comment:
there isn't a running clock so the OR checks to see if the next sleep duration would exceed the timeout rather than go to sleep unnecessarily (and hit the proxy timeout).